File: des--elev_rud_model_cpp_ws/src/auv_imp/src/auv_imp.cpp
#include <ros/ros.h>
#include <avl_bsd_translator/CommandMsg.h>
#include <avl_bsd_translator/ActionMsg.h>
#include <avl_bsd_translator/HeartbeatMsg.h>
#include <std_msgs/Float64.h>
#include <math.h>

class AUVNode {
public:
    AUVNode() {
        // Initialize ROS node
        ros::NodeHandle nh;

        // Initialize control variables
        elevator = 0.0;
        rudder = 0.0;
        rpm = 0.0;
        z = 250.0;

        // Set publishing frequency
        ros::Rate rate(1);

        // Create publisher for heartbeat messages
        heartbeat_pub = nh.advertise<avl_bsd_translator::HeartbeatMsg>("bsd/heartbeat", 10);

        // Create publisher for z topic
        z_pub = nh.advertise<std_msgs::Float64>("bsd/z", 10);

        // Create subscribers for command and action topics
        command_sub = nh.subscribe("bsd/command", 10, &AUVNode::command_callback, this);
        action_sub = nh.subscribe("bsd/action", 10, &AUVNode::action_callback, this);

        update_heartbeat();
    }

    void command_callback(const avl_bsd_translator::CommandMsg::ConstPtr& msg) {
        ROS_INFO("Command message received");
    }

    void action_callback(const avl_bsd_translator::ActionMsg::ConstPtr& msg) {
        elevator = std::stod(msg->VALUES[1]);
        rudder = std::stod(msg->VALUES[2]);
        rpm = std::stod(msg->VALUES[3]);

        ROS_INFO("Action message received");
    }

    void update_heartbeat() {
        double roll = 0.0;
        double pitch = 0.0;
        double yaw = 0.0;
        double x = 0.0;
        double y = 0.0;
        double z = this->z;
        double u = this->elevator;
        double v = this->rudder;
        double w = 0.0;
        double p = 0.0;
        double q = 0.0;
        double r = 0.0;

        double dt = 1.0;

        double elevator_rad = M_PI / 180.0 * this->elevator;
        double rudder_rad = M_PI / 180.0 * this->rudder;

        double depth = z + (w * dt) + (0.5 * elevator_rad * pow(dt, 2));
        x += u * dt;
        y += v * dt;
        this->z = depth;

        avl_bsd_translator::HeartbeatMsg heartbeat_msg;
        heartbeat_msg.HEARTBEAT_STATUS = 1;
        heartbeat_msg.HEARTBEAT_ROLL = roll;
        heartbeat_msg.HEARTBEAT_PITCH = pitch;
        heartbeat_msg.HEARTBEAT_YAW = yaw;
        heartbeat_msg.HEARTBEAT_VN = u;
        heartbeat_msg.HEARTBEAT_VE = v;
        heartbeat_msg.HEARTBEAT_VD = w;
        heartbeat_msg.HEARTBEAT_DEPTH = this->z;
        heartbeat_msg.HEARTBEAT_HEIGHT = y;
        heartbeat_msg.HEARTBEAT_RPM = this->rpm;
        heartbeat_msg.HEARTBEAT_WX = p;
        heartbeat_msg.HEARTBEAT_WY = q;
        heartbeat_msg.HEARTBEAT_WZ = r;
        heartbeat_msg.HEARTBEAT_AX= 0.0;
        heartbeat_msg.HEARTBEAT_AY = 0.0;
        heartbeat_msg.HEARTBEAT_AZ = 0.0;
        heartbeat_msg.HEARTBEAT_VOLTAGE = 0.0;
        heartbeat_msg.HEARTBEAT_UMODEM_SYNC = 0;
        heartbeat_msg.HEARTBEAT_MISSION_MODE = 0;
        heartbeat_msg.HEARTBEAT_FSD_MISSION_STATE = 0;
        heartbeat_msg.HEARTBEAT_FSD_CURRENT_ACTION = 0;
        heartbeat_msg.HEARTBEAT_FSD_TOTAL_ACTIONS = 0;
        heartbeat_msg.HEARTBEAT_FSD_ACTION_PERCENT = 0.0;
        heartbeat_msg.HEARTBEAT_BSD_MISSION_STATE = 0;
        heartbeat_msg.HEARTBEAT_BSD_CURRENT_ACTION = 0;
        heartbeat_msg.HEARTBEAT_BSD_TOTAL_ACTIONS = 0;
        heartbeat_msg.HEARTBEAT_BSD_ACTION_PERCENT = 0.0;
        heartbeat_msg.HEARTBEAT_GPS_SATS = 0;
        heartbeat_msg.HEARTBEAT_GPS_LAT = 0.0;
        heartbeat_msg.HEARTBEAT_GPS_LON = 0.0;
        heartbeat_msg.HEARTBEAT_GPS_ALT = 0.0;
        heartbeat_msg.HEARTBEAT_NAV_LAT = 0.0;
        heartbeat_msg.HEARTBEAT_NAV_LON = 0.0;
        heartbeat_msg.HEARTBEAT_NAV_ALT = 0.0;
        heartbeat_msg.HEARTBEAT_NAV_YAW_STD = 0;
        heartbeat_msg.HEARTBEAT_NAV_AVG_POS_ERR = 0;
        heartbeat_msg.HEARTBEAT_NAV_INITIALIZED = 0;

        // Check if the publisher is valid before publishing
        if (heartbeat_pub.getNumSubscribers() > 0 && ros::ok()) {
            heartbeat_pub.publish(heartbeat_msg);
        }

        // Publish z value
        std_msgs::Float64 z_msg;
        z_msg.data = this->z;
        z_pub.publish(z_msg);
    }

private:
    double elevator;
    double rudder;
    double rpm;
    double z;
    ros::Publisher heartbeat_pub;
    ros::Publisher z_pub;
    ros::Subscriber command_sub;
    ros::Subscriber action_sub;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "auv_node");

    AUVNode node;

    ros::spin();

    return 0;
}

File: des--elev_rud/src/auv_imp/nodes/auv_imp.py
#!/usr/bin/env python3

import rospy
import math

from avl_bsd_translator.msg import CommandMsg, ActionMsg, HeartbeatMsg

class HeartbeatPublisher:

    def __init__(self):
        rospy.init_node('auv_imp')

        self.heartbeat_pub = rospy.Publisher('bsd/heartbeat', HeartbeatMsg, queue_size=10)
        self.command_sub = rospy.Subscriber('bsd/command', CommandMsg, self.command_callback)
        self.action_sub = rospy.Subscriber('bsd/action', ActionMsg, self.action_callback)

        self.heartbeat_msg = HeartbeatMsg()
        self.heartbeat_msg.HEARTBEAT_STATUS = 1
        self.heartbeat_msg.HEARTBEAT_ROLL = 2.0
        self.heartbeat_msg.HEARTBEAT_PITCH = 2.0
        self.heartbeat_msg.HEARTBEAT_YAW = 2.0
        self.heartbeat_msg.HEARTBEAT_VN = 0.0 # VX origin
        self.heartbeat_msg.HEARTBEAT_VE = 0.0 # VY origin
        self.heartbeat_msg.HEARTBEAT_VD = 0.0 # VZ origin
        self.heartbeat_msg.HEARTBEAT_DEPTH = 250.0
        self.heartbeat_msg.HEARTBEAT_HEIGHT = 0.0
        self.heartbeat_msg.HEARTBEAT_RPM = 1500.0
        self.heartbeat_msg.HEARTBEAT_WX = 0.0
        self.heartbeat_msg.HEARTBEAT_WY = 0.0
        self.heartbeat_msg.HEARTBEAT_WZ = 0.0
        self.heartbeat_msg.HEARTBEAT_AX = 0.0
        self.heartbeat_msg.HEARTBEAT_AY = 0.0
        self.heartbeat_msg.HEARTBEAT_AZ = 0.0
        self.heartbeat_msg.HEARTBEAT_VOLTAGE = 0.0
        self.heartbeat_msg.HEARTBEAT_UMODEM_SYNC = 1
        #self.heartbeat_msg.HEARTBEAT_IRIDIUM_STRENGTH = 4
        self.heartbeat_msg.HEARTBEAT_MISSION_MODE = 3
        self.heartbeat_msg.HEARTBEAT_FSD_MISSION_STATE = 0
        self.heartbeat_msg.HEARTBEAT_FSD_CURRENT_ACTION = 1
        self.heartbeat_msg.HEARTBEAT_FSD_TOTAL_ACTIONS = 5
        self.heartbeat_msg.HEARTBEAT_FSD_ACTION_PERCENT = 20.0
        self.heartbeat_msg.HEARTBEAT_BSD_MISSION_STATE = 0
        self.heartbeat_msg.HEARTBEAT_BSD_CURRENT_ACTION = 2
        self.heartbeat_msg.HEARTBEAT_BSD_TOTAL_ACTIONS = 3
        self.heartbeat_msg.HEARTBEAT_BSD_ACTION_PERCENT = 50.0
        self.heartbeat_msg.HEARTBEAT_GPS_SATS = 8
        self.heartbeat_msg.HEARTBEAT_GPS_LAT = 37.7749
        self.heartbeat_msg.HEARTBEAT_GPS_LON = -122.4194
        self.heartbeat_msg.HEARTBEAT_GPS_ALT = 0.0
        self.heartbeat_msg.HEARTBEAT_NAV_LAT = 37.7749
        self.heartbeat_msg.HEARTBEAT_NAV_LON = -122.4194
        self.heartbeat_msg.HEARTBEAT_NAV_ALT = 0.0
        self.heartbeat_msg.HEARTBEAT_NAV_YAW_STD = 0
        self.heartbeat_msg.HEARTBEAT_NAV_AVG_POS_ERR = 0
        self.heartbeat_msg.HEARTBEAT_NAV_INITIALIZED = 1

    def send_heartbeat_msg(self, event):
        self.heartbeat_pub.publish(self.heartbeat_msg)

    def command_callback(self, msg):
        time = rospy.get_time()
        fluctuation1 = 1.0 * math.sin(0.1 * time)  # Amplitude of 1 with frequency of 0.1
#        self.heartbeat_msg.HEARTBEAT_ROLL = 2.0 + fluctuation1
        self.heartbeat_msg.HEARTBEAT_PITCH = 2.0 + fluctuation1
        self.heartbeat_msg.HEARTBEAT_YAW = 2.0 + fluctuation1
        self.heartbeat_msg.HEARTBEAT_ROLL = 2.0
#        self.heartbeat_msg.HEARTBEAT_PITCH = 2.0
#        self.heartbeat_msg.HEARTBEAT_YAW = 2.0
    def action_callback(self, msg):
        time = rospy.get_time()
        fluctuation = 1.0 * math.sin(0.1 * time)  # Amplitude of 1 with frequency of 0.1
        self.heartbeat_msg.HEARTBEAT_DEPTH = 250.0 + fluctuation
#        self.heartbeat_msg.HEARTBEAT_DEPTH = 250.0
    def run(self):
        rospy.Timer(rospy.Duration(0.1), self.send_heartbeat_msg)
        rospy.spin()

if __name__ == '__main__':
    heartbeat_publisher = HeartbeatPublisher()
    heartbeat_publisher.run()
